src/Shared/PackageManagers/MavenProjectManager.cs

```diff
@@ -315,48 +459,67 @@ private static MavenArtifactType GetMavenArtifactType(string fileName)
-            if (string.IsNullOrEmpty(html))
+
+            if (upstream == MavenSupportedUpstream.MavenCentralRepository)
             {
-                throw new InvalidOperationException();
-            }
+                var baseUrl = $"{upstream.GetRepositoryUrl().EnsureTrailingSlash()}{packageNamespace}/{packageName}/";
 
-            HtmlParser parser = new();
-            AngleSharp.Html.Dom.IHtmlDocument document = await parser.ParseDocumentAsync(html);
+                string? html = await GetHttpStringCache(httpClient, baseUrl, useCache);
+                if (string.IsNullOrEmpty(html))
+                {
+                    throw new InvalidOperationException();
+                }
 
-            // Break the version content down into its individual lines, and then find the one that represents the
-            // intended version.
-            string? versionContent = document.QuerySelector("#contents").TextContent
-                .Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
-                .SingleOrDefault(versionText => versionText.StartsWith($"{packageVersion}/"));
+                HtmlParser parser = new();
+                AngleSharp.Html.Dom.IHtmlDocument document = await parser.ParseDocumentAsync(html);
 
-            if (versionContent == null)
-            {
-                return null;
-            }
+                // Break the version content down into its individual lines, and then find the one that represents the
+                // intended version.
+                string? versionContent = document.QuerySelector("#contents").TextContent
+                    .Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
+                    .SingleOrDefault(versionText => versionText.StartsWith($"{packageVersion}/"));
+
+                if (versionContent == null)
+                {
+                    return null;
+                }
 
-            // Split the version content into its individual parts.
-            // [0] - The version
-            // [1] - The date it was published
-            // [2] - The time it waas published
-            // [3] - The download count
-            string[] versionParts = versionContent.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
-            if (DateTime.TryParse($"{versionParts[1]} {versionParts[2]}", out DateTime publishDateTime))
+                // Split the version content into its individual parts.
+                // [0] - The version
+                // [1] - The date it was published
+                // [2] - The time it waas published
+                // [3] - The download count
+                string[] versionParts = versionContent.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
+                if (DateTime.TryParse($"{versionParts[1]} {versionParts[2]}", out DateTime publishDateTime))
+                {
+                    return publishDateTime;
+                }
+            }
```

Comment:
I've seen a couple places where we fetch the directory index. Should we extract this to a single implementation that we can reuse?

Reply:
I kept it separate for now because there's only one shared line that extracts the index. Any content parsing is unique based on what we're looking for.

Reply:
Hmm. I hear your point about the different methods selecting different things from the HTML. I'm not attached to consolidating these, but I can't help but wonder if we could benefit from having fewer places we need to think about error handling, debugging, etc. if we were to have one implementation of how to parse the directory index HTML that could fetch everything we might want. Something that could spit out like a list of file name & publish date time pairs, or a dictionary from file name to publish date time, etc.

Reply:
Yeah I can try the dictionary approach. FYI not all repositories that follow the directory index approach provide the published timestamp (e.g. [https://plugins.gradle.org/m2/Hoop/conductor-lifecycle/0.0.1/](https://plugins.gradle.org/m2/Hoop/conductor-lifecycle/0.0.1/)). But I can see the benefit of having one method for filename to publish datetime correlation.